### PR TITLE
Fix `thread_mattr_accessor` `default` option behavior

### LINKED ISSUE #43312: `thread_mattr_accessor` default value does not apply to threads
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "active_support/core_ext/module/attribute_accessors_per_thread"
require "minitest/autorun"

class BugTest < Minitest::Test
  class ThreadAccessors
    thread_mattr_accessor :variable, default: 42
  end

  def test_stuff
    assert(ThreadAccessors.variable == 42)
    Thread.new do
      assert(ThreadAccessors.variable == 42)
    end.join
  end
end
```

### Expected behavior

I'd expect the default value for the thread variable to be present in all threads

### Actual behavior

The variable is initialized only in the main thread at load time and `nil` in other threads

### System configuration
**Rails version**: main branch

**Ruby version**: ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]

### PR DESCRIPTION
**Pull request by tdeo**
### Summary

This makes the value supplier to the `default` option of
`thread_mattr_accessor` to be set in descendant classes
as well as in any new Thread that starts.

Previously, the `default` value provided was set only at the
moment of defining the attribute writer, which would cause
the attribute to be uninitialized in descendants and in other threads.

For instance:

    class Processor
      thread_mattr_accessor :mode, default: :smart
    end

    class SubProcessor < Processor
    end

    SubProcessor.mode # => :smart
    Thread.new do
      Processor.mode # => :smart
    end.join

when in the past, those two calls would have returned `nil`.

Fixes #43312

### Other Information

This logic comes with a performance impact on the reader,
here is the benchmark:

    Warming up --------------------------------------
            original   238.780k i/100ms
             patched   162.765k i/100ms
    Calculating -------------------------------------
            original    2.376M (± 9.4%) i/s -  11.939M in 5.078355s
             patched    1.635M (±16.4%) i/s -   7.813M in 5.008433s

    Comparison:
            original:  2375953.1 i/s
             patched:  1634668.2 i/s - 1.45x  (± 0.00) slower

### CODE REVIEW

**Comment by tdeo** [comment]
@ghiculescu do you know if there are any blockers on this topic? Do we rather expect people to implement this behaviour in user-land?

File: activesupport/CHANGELOG.md [hunk 1]
```diff
@@ -5,4 +5,15 @@
 
     *Jean Boussier*
 
+*   `default` option of `thread_mattr_accessor` now applies through inheritance and
```
Comment (ghiculescu) [inline activesupport/CHANGELOG.md:8]: Could you rebase and then move this to the top of the changelog?

**Comment by ghiculescu** [comment]
It seems fine to me, but we need someone from the core team to review. They will get to it, but it can take some time.

**Review by jonathanhefner — COMMENTED** [review]
I played around with the code to see if I could reduce the performance impact on the reader.  I was able to limit the performance impact to ~7% in most cases.  However, in the case of setting an attribute to `nil` when a non-`nil` default is declared, the impact is ~45%.  Even in that case, though, the code still runs at > 1M ops/sec, so it may be an acceptable trade-off.

<details>
  <summary><strong>Benchmark script</strong></summary>

  ```ruby
  # frozen_string_literal: true
  require "benchmark/ips"
  require "active_support"
  require "active_support/core_ext/module/attribute_accessors_per_thread"

  class MyClass
    thread_mattr_accessor :default_value, default: "default"
    thread_mattr_accessor :string_value, default: "default"
    thread_mattr_accessor :nil_value, default: "default"
  end

  MyClass.string_value = "string"
  MyClass.nil_value = nil

  Benchmark.ips do |x|
    x.report("default_value") { MyClass.default_value }
    x.report("string_value") { MyClass.string_value }
    x.report("nil_value") { MyClass.nil_value }
  end
  ```
</details>

`main` branch results:

  ```
         default_value      2.075M (± 0.7%) i/s -     10.396M in   5.010585s
          string_value      2.103M (± 0.7%) i/s -     10.672M in   5.074624s
             nil_value      1.777M (± 0.9%) i/s -      8.924M in   5.023058s
  ```

This branch results:

  ```
         default_value      2.008M (± 0.7%) i/s -     10.187M in   5.072990s
          string_value      1.967M (± 0.7%) i/s -      9.891M in   5.028570s
             nil_value      1.144M (± 0.5%) i/s -      5.770M in   5.041630s
  ```

(When no default / `nil` default is specified, the performance will be the same as `main`.)

I also switched away from a class variable to fix the case of a subclass overriding the default value of a superclass.  (See the [`test_subclass_can_override_default_value_without_affecting_superclass`](https://github.com/rails/rails/pull/43337/files#diff-ba91a8112fee46e4dbba5fd31262773420b027a778097571f5115586c66cfe9dR202-R208) test.)

I went ahead and pushed the commit (5ee887853aa09f0c207d2fefdb820272a6177e3c) to this PR.  Let me know what you think.

**Comment by jonathanhefner** [comment]
The build had [*lots* of failures](https://buildkite.com/rails/rails/builds/89015) that I could not reproduce locally, so I rebased on `main` and force-pushed.  I hope you do not mind. :pray:

**Comment by tdeo** [comment]
Hey @jonathanhefner thanks for looking at this benchmark more thoroughly and for the additional tests, changes look great.

**Comment by jonathanhefner** [comment]
Great!  I went ahead and squashed the commits.  Thank you, @tdeo! :tada:

**Comment by ngan** [comment]
Hi, we got bitten by this bug and were looking to backport this fix. But as I'm looking at the code, it looks like the `default` value is assigned when the value is fetched for the first time in a thread. But if the `default` value is, say, a hash (`default: {}`), would that mean that the exact same hash is shared by all threads? A mutation to the hash in one thread would leak over to the others.

For some reason, I recall that `thread_mattr_accessor` took a `Proc` as the default value so that it could be evaluated at lookup. That way you can do:

```ruby
thread_mattr_accessor :foo, default: -> { Hash.new }
```

to get a new/different hash for every thread. I could be wrong.

**Comment by ngan** [comment]
Ah it was here...at the first implementation. `thread_mattr_accessor` took a block for its default value:
https://github.com/rails/rails/commit/2bd2d853da78c248b1d50add92ce21f43430abe9#diff-69bbb51e29b9d745fb73e838fa9cd730fc029786ed581e1e4e39ee95090f7660R56-R57

So you can do this:
```ruby
thread_mattr_reader :foo do
  Hash.new
end
```

**Comment by jonathanhefner** [comment]
> But if the `default` value is, say, a hash (`default: {}`), would that mean that the exact same hash is shared by all threads? A mutation to the hash in one thread would leak over to the others.

Indeed!  So as a follow-up, I submitted #45918 to provide a basic level of protection against that.

**Pull request merged**